### PR TITLE
Support listening both IPv4 and IPv6 + NodeRecord extension

### LINKED ISSUE #176: Support listening and advertising dual-stack (IPv4 and IPv6)
**Description**

In order to support dual-stack (IPv4 and IPv6), discovery should be able to configure multiple listening addresses and advertising dual-stack ENR.

### PR DESCRIPTION
**Pull request by StefanBratanov**
## PR Description

- Allow to pass multiple listen addresses and ports
- Expose `IPv6` methods in `NodeRecord`

Change CircleCI to use machine executor instead of docker as per https://circleci.com/docs/faq/#ipv6-in-tests

## Fixed Issue(s)
fixes  #176

### CODE REVIEW

**Review by zilm13 — APPROVED** [review]
LGTM

File: src/main/java/org/ethereum/beacon/discovery/network/NettyDiscoveryServerImpl.java [hunk 1]
```diff
@@ -45,7 +45,8 @@ public CompletableFuture<NioDatagramChannel> start() {
     LOG.info("Starting discovery server on UDP port {}", listenAddress.getPort());
     if (!listen.compareAndSet(false, true)) {
       return CompletableFuture.failedFuture(
-          new IllegalStateException("Attempted to start an already started server"));
+          new IllegalStateException(
+              "Attempted to start an already started server on port " + listenAddress.getPort()));
```
[the remark this answers is not in the document]
Comment (StefanBratanov) [reply src/main/java/org/ethereum/beacon/discovery/network/NettyDiscoveryServerImpl.java]: Yeah added that

File: src/main/java/org/ethereum/beacon/discovery/network/NettyDiscoveryClientImpl.java [hunk 2]
```diff
@@ -17,17 +20,20 @@
...
-   * @param channel Nio channel
+   * @param channels must be either 1 (IPv4/IPv6) or 2 (IPv4 and IPv6)
    */
   public NettyDiscoveryClientImpl(
-      Publisher<NetworkParcel> outgoingStream, NioDatagramChannel channel) {
-    this.channel = channel;
+      final Publisher<NetworkParcel> outgoingStream, final List<NioDatagramChannel> channels) {
+    this.ip4Channel = channels.stream().filter(channel -> !isChannelIPv6(channel)).findFirst();
+    this.ip6Channel = channels.stream().filter(this::isChannelIPv6).findFirst();
     Flux.from(outgoingStream)
```
[the remark this answers is not in the document]
Comment (StefanBratanov) [reply src/main/java/org/ethereum/beacon/discovery/network/NettyDiscoveryClientImpl.java:35]: Changed the design a bit. You can have a look.

**Review by Nashatyrev — APPROVED** [review]
LGTM

File: src/main/java/org/ethereum/beacon/discovery/network/NettyDiscoveryClientImpl.java [hunk 3]
```diff
@@ -17,17 +19,19 @@
...
    * @param outgoingStream Stream of outgoing packets, client will forward them to the channel
-   * @param channel Nio channel
+   * @param channels must have either 1 entry (IPv4/IPv6) or 2 entries (IPv4 and IPv6)
    */
   public NettyDiscoveryClientImpl(
-      Publisher<NetworkParcel> outgoingStream, NioDatagramChannel channel) {
-    this.channel = channel;
+      final Publisher<NetworkParcel> outgoingStream,
+      final Map<InternetProtocolFamily, NioDatagramChannel> channels) {
+    this.channels = channels;
```
Comment (zilm13) [inline src/main/java/org/ethereum/beacon/discovery/network/NettyDiscoveryClientImpl.java:34]: I'd add throw if channels is empty.

**Pull request merged**